### ISSUE TITLE
BF 3.4 + 3.5 Nightly, R-XSR (inverted) on FPORT, lua v1.0.2 - runs 'sub-optimally' :-)

### ISSUE THREAD

**Issue by RipperDrone** [betaflight/betaflight-tx-lua-scripts#151]
All kinds of glitches:

- sensor lost messages quite freqently on Taranis X9D+(even though I have no ESC telemetry enabled, no telemetry sensors like ESC, fuel, AccX/Y/Z, GPS set up
- some lua pages loading values, others don't
- some values saving, others don't
- generally annoying 'retrying' message every time, even though sometimes values are actually saved

Taranis FW is 2.2.2 RC5, lua scripts running somewhat more stable on BF3.4 than on 3.5... :-/

**Comment by mikeller** [comment]
What firmware do you have on your RX?

Also, providing information about the versions, and a config `diff`, like you were asked to do when opening this pull request, will greatly improve the chances of anything being done about your rant.

**Comment by RipperDrone** [comment]
No intention to make it understood as a rant / see smilie in title.

I didn't provide all the details since I thought it might be a general issue (also mentioned in some other issue reports before) that the latest lua script still does not cover all the new timings of BFf3.4+ regarding FPORT protocol.

Anyway, here's the details you asked for:
- RX Firmware version is 180228, latest stable from FrSky webpage
- BF 3.4 stable, BF 3.5 Nightly #1010 tested
- diff link: https://drive.google.com/open?id=1PD-H_9OGwnP7vYVieSq3KS48oPFjYnU_

**Comment by mikeller** [comment]
@RipperDrone: For some reason I get a 404 when trying to open your diff. Can you try 3.5.0-RC1? betaflight/betaflight#6397 disables sending of the per-ESC telemetry data, reducing the bandwidth going through telemetry.

**Comment by RipperDrone** [comment]
Link updated. Will test 3.5RC1 next...

**Comment by RipperDrone** [comment]
mixed results:

Quad1 = OBF4: All telemetry fine with BF 3.4 stable
Quad2 = DYSF4PRO: BF3.4 = 2 missing tele pages Rates and motor PWM. BF3.5RC1: No telemetry pages at all with lua script.

**Comment by RipperDrone** [comment]
Flashed Quad2 back to BF3.4 - again script working but for 2 pages (no values shown, others are fine)...

**Comment by Guidus93** [comment]
Also here, no values on Lua with 3. 5,it was working with 3.3 and 3.4 TOO. RXSR with stock eu firmware and Fortini 
Is there something I should check on cli parameters?

**Comment by RipperDrone** [comment]
must have to do with altered tele frames + timings...?

**Comment by klutvott123** [comment]
With 3.5.0 RC1 i stopped getting "sensor lost" all the time, but as soon as i switch to the telemetry page with the lua script i get "sensor lost" after a few seconds. I don't get any values on any pages of the lua script either. If i don't switch to the lua script everything is fine and telemetry works perfectly. 

Tried 3.3.3 again just to test and it shows values on most of the lua pages( not PWM and VTX). I can change and save values, but if i do it too much it's "sensor lost" again. With 3.3.3 i have to wait for a few minutes for the sensor to come back. Rebooting FC/transmitter does nothing. "sensor lost" seconds after switching on. 

I'm starting to suspect that the issue is with the R-XSR as it seems like most people with these issues are using it. Those using other receivers had issues with ESC telemetry enabled. 
Also @mikeller mentioned in #120 that the R-XSR has a 8051 microcontroller instead of the stm32 that the other frsky receivers are using and suggests that it might not be capable of handling the same number of sensors. 

I might get a X4RSB to see if there's any difference. 

Using SPRACINGF3 clone, R-XSR  f.port 180615 beta firmware(tried all firmwares),
Taranis q x7 with opentx 2.2.2 RC5.

**Comment by RipperDrone** [comment]
R-XSR FPORT was spot on with BF3.4 (had all the probs you described before thqat) on many FCs like OBF4, OBF4SD. On DYSF4PRO,  2 telemetry pages over lua were not displaying values, all the others are ok. So since firmware version 180228, the R-XSR seemed to work almost by 100%. But then, with BF3.5RC1, none of the values / pages was working anymore.

It's true the R-XSR has a different MCU, and @mikeller has done a great job in adjusting frames frequency, timing and so forth. But it still seems there is some timing glitch in BF3.5 now...

**Comment by mikeller** [comment]
@RipperDrone: Unfortunately, given we have got SmartPort and FPort, a number of different RX, all with a number of firmware versions, 'making it work' is more like 'herding cats'...

Can you try the 180615 version for r-XSR (under 'Beta firmware for FPort)?

**Comment by RipperDrone** [comment]
Yes, I know. Will update the OBF4 where RX is easier to access. On my Airdancer with the DYSF4Pro, I would have to tear a lot of soldering and shrinks apart since the RX is buried deep underneath the center FC section which is tied down / constrained by short ESC cables and shrik wraps around the arms. Extremely painful, trying to avoid this. When on a promising track, I can test with other rigs, though...

It seems that the worst combination right now is the DYSF4PRO / R-XSR FPORT, where I don't get ANY of the lua telemetry pages filled with readout values. This rig has BLHeli32 individual ESCs which issue telemetry data. I have tied those 4 telemetry pins together and attached them to one UART.

**Comment by RipperDrone** [comment]
Did some testing now, rather frustrating than enlightening... :-)

- OBF4 quad didn't show any changes after flashing latest FrSky FPORT Beta (180615). I then somehow must have kiled the gyro of the HGLRC stack (really bad quality soldering they do), so I had to replace the FC and solder everything new. Connected SBUS + separate telemetry this time. Now BF 3.4 and SBUS latest FrSky firmware on R-XSR working well. Cutting FPORT out of the equation did help here. Of course, it doesn't help the team here to fix FPORT, though...
- other FC DYSF4PRO (which I can't access the RX easily w/o destroying the whole build basically :-(): I tried BF3.4 and 3.5 with follwing swaps: Telemetry feature ON/OFF, ESC sensor feature ON/OFF, Port settings including ESC: ON/OFF, serial settings inverted / halfduplex all combinations. In any of these cases, lua script 1.0.2 works well on BF3.4 except the 2 pages rates and motor PWM. On BF3.5 #1033 NO page is being displayed with any values, completely dysfunctional. There must be something fundamentally changed btw BF3.4 and 3.5 re lua telemetry redout timings / frames...

**Comment by RipperDrone** [comment]
to try flash latest R-XSR FPORT beta firmware w/o rebuilding the whole quad: Is there a way to flash the RX over serial passthrough in the meantime so I could stay wired up as it is?

**Comment by RipperDrone** [comment]
found something: Rate page hasbeen empty in BF3.3 as well under certain circumstances... https://github.com/betaflight/betaflight-tx-lua-scripts/issues/117

**Comment by RipperDrone** [comment]
Could it be PR https://github.com/betaflight/betaflight/pull/6300 causing the difference BF3.4 <-> 3.5 maybe?

**Comment by klutvott123** [comment]
Changing FPORT_MIN_TELEMETRY_RESPONSE_DELAY_US to 100 in fport.c got rid of the "sensor lost" for me when opening the lua script. Still no values on any pages. Then i increased FPORT_TELEMETRY_MAX_CONSECUTIVE_TELEMETRY_FRAMES to 10 and now i get values on all pages except PWM and VTX(don't know if i'm supposed to have vtx values since i don't have a vtx.). 
Still a little slow. Takes a few seconds to load pages but i can change values and save them. Just flew 4 packs and did a little pid tuning with no problems other than one quick "sensor lost". 
Works for my setup but might not for other setups. I can see how getting this to work for every setup is like "herding cats". 

Also compiled 3.5.0 with fport.c from 3.3 and that also works(still no PWM and VTX). I think it's because 3.3 didn't have the restriction on consecutive telemetry frames at all.

**Comment by mikeller** [comment]
@klutvott123: `FPORT_MIN_TELEMETRY_RESPONSE_DELAY_US` is something that was directly specified by FrSky, so making it shorter will potentially lead to problems with future versions of the RX firmware, since that's the spec they make it for.

`FPORT_TELEMETRY_MAX_CONSECUTIVE_TELEMETRY_FRAMES` had to be introduced to prevent buffer overruns on r-XSR. This is superseded by the change introduced in betaflight/betaflight#6098 that allows the RX to signal to the flight controller when it's ready to get another telemetry frame. But this requires RX firmware support. Currently the r-XSR 'Beta firmware for FPort' supports this, and there is XSR firmware in the works for it.

**Comment by klutvott123** [comment]
@mikeller Thanks for explaining. I just reflashed my R-XSR with the latest 180615 beta fport firmware just to be sure and now it's working without changing FPORT_MIN_TELEMETRY_RESPONSE_DELAY_US and FPORT_TELEMETRY_MAX_CONSECUTIVE_TELEMETRY_FRAMES. I was sure i was on 180615 but must have had 180228 instead(have been switching between the two to test). 

The PWM and VTX pages are still empty. Do you know what might be causing this?

**Comment by RipperDrone** [comment]
Weird thought crossed my mind: Does anyone of you who is having the arms drop problem have pid_at_min_throttle switched off? It's off on all my quads since I prefer to have the quad sitting idle on launch, no slow I term rpm windups whatsoever, no reaction to stick inputs unless I raise throttle.

I understand that this switch shouldn't matter since once a certain throttle threshold is surpassed, PID control should engage and stay on from then, so even hard throttle cuts back to min throttle should be PID covered. Confirmed? (I will try to swap the setting of this option, anyway... 🙂)

**Comment by RRAlex** [comment]
Same issue here on OpenTX 2.2.2 and BF 3.4.1 with F.Port 180228:

Lost telemetry & slow update (though I can't compare, this is my first build...)

**Comment by RRAlex** [comment]
@klutvott123 Oh, I see, didn't realize there was a beta, will try that and give some feedbacks, thanks!

**Comment by atomiix** [comment]
Same issue here on OpenTX 2.2.2, BF 3.5.0 and X4RSB with F.Port 171114 (last nonEU version).
Some "sensor lost", slow update & empty PWM & VTX.

**Comment by klutvott123** [comment]
In https://github.com/betaflight/betaflight/blob/master/src/main/interface/msp.c line 1240:

`    case MSP_ADVANCED_CONFIG:
        sbufWriteU8(dst, gyroConfig()->gyro_sync_denom);
        sbufWriteU8(dst, pidConfig()->pid_process_denom);
        sbufWriteU8(dst, motorConfig()->dev.useUnsyncedPwm);
        sbufWriteU8(dst, motorConfig()->dev.motorPwmProtocol);
        sbufWriteU16(dst, motorConfig()->dev.motorPwmRate);
        sbufWriteU16(dst, motorConfig()->digitalIdleOffsetValue);
        sbufWriteU8(dst, gyroConfig()->gyro_use_32khz);
        sbufWriteU8(dst, motorConfig()->dev.motorPwmInversion);
        sbufWriteU8(dst, gyroConfig()->gyro_to_use);
        sbufWriteU8(dst, gyroConfig()->gyro_high_fsr);
        sbufWriteU8(dst, gyroConfig()->gyroMovementCalibrationThreshold);
        sbufWriteU16(dst, gyroConfig()->gyroCalibrationDuration);
        sbufWriteU16(dst, gyroConfig()->gyro_offset_yaw);
sbufWriteU8(dst, gyroConfig()->checkOverflow);`

The bottom 6 lines are new from 3.4. I tried commenting them out just to test and that got the PWM working again. I can't see why it's not working with the above code. The FC just sends all those bytes and the lua script uses the first 9.  I tried sending the MSP_ADVANCED_CONFIG command in a terminal to see what exactly comes out but didn't have any luck with that.

**Comment by atomiix** [comment]
@klutvott123 I can confirm, I went further and tried to comment each line one by one and it's the line `sbufWriteU16(dst, gyroConfig()->gyroCalibrationDuration);` which cause the issue.

So, I did `set gyro_calib_duration = 128` in the CLI and the PWM page was working again !

**Comment by klutvott123** [comment]
@atomiix I just set it to 128 too and it's working now. Thanks!

This has to be some kind of bug because i have something like this happening on the PID page too. Some combinations of pid gains just makes the page blank. Right now if i set the pitch I to 60 and roll I to 55 it stops working. Earlier i had a case where setting yaw I to 55 it stopped working. I could set it to 54 or 56 and all other values but not 55. 

Same thing also happens when setting the d-term notch filter center frequency to 125 and the cutoff to 75.

**Comment by atomiix** [comment]
@klutvott123 I had exactly the same problem when I was tuning my PIDs, that's what led me to try another value for `gyro_calib_duration`

**Comment by atomiix** [comment]
It seems that under certain circumstances, F.Port frames are dropping packets.

I got the line `elseif bit32.band(mspRemoteSeq + 1, 0x0F) ~= seq then` of the file [src/SCRIPTS/BF/MSP/common.lua](https://github.com/betaflight/betaflight-tx-lua-scripts/blob/master/src/SCRIPTS/BF/MSP/common.lua#L105) (line105) being called with the page PWM (which means packet drop).
It occures only on F.Port, I tried back with SBUS/SmartPort and it's working well.

But I'm not competent enough to investigate more on the F.Port protocol 😅

**Comment by mikeller** [comment]
@atomiix: There are a number of places where telemetry frames can be dropped:
- due to having only a 2 bit sequence number on the radio link, any loss of 3 or more frames will result in a data loss on FPort and SmartPort;
- in FPort, the initial implementation had the flight controller send more telemetry data to the RX than the RX could send over the air, causing queue overruns. This has been fixed in the latest Betaflight versions, but the fix so far only exists for R-XSR on the RX side ('Beta' firmware).

**Comment by klutvott123** [comment]
@mikeller I just read the f.port protocol documentation and did some experimenting. If i set any value on any lua script page to 125 or 126 it stops working. 
126 is the FPORT_FRAME_MARKER 0x7E and 125 is the FPORT_ESCAPE_CHAR 0x7D in https://github.com/betaflight/betaflight/blob/master/src/main/rx/fport.c. This makes me believe the f.port bytestuffing isn't properly implemented somewhere. 

https://github.com/betaflight/betaflight/blob/98616d469012e5287f35c9302bbea70d1e8b920d/src/main/telemetry/smartport.c#L274 This looks right to me so i assume the f.port frame is correct when sent to the receiver. Is there a chance that it's not properly "unstuffed" at the receiver side and not getting sent to the transmitter at all when the data contains 125 or 126?

Using betaflight 3.5, R-XSR with 180615 beta f.port firmware.

**Comment by klutvott123** [comment]
Did some more testing. I set my PIDs so that the PID page stopped working. I wrote a python script that sends the MSP_PID command to the FC and prints the received data. The combination of pid gains that makes it stop working has a CRC of 125(FPORT_ESCAPE_CHAR 0x7D). Now i really think this is a bytestuffing issue.

**Comment by mikeller** [comment]
Good find, @klutvott123!

On the flight controller side, FPort uses the same byte stuffing that SmartPort uses: https://github.com/betaflight/betaflight/blob/98616d469012e5287f35c9302bbea70d1e8b920d/src/main/rx/fport.c#L237-L239

But where I can see a potential difference is that, because with FPort, the frame delimiters are used for RC control and telemetry frames on the same connection, any occurrence of a stray delimiter would have way worse consequences than on SmartPort, where the RC control is unaffected.

My suspicion is that the lua API that we are using to inject the MSP data expects the data to be byte stuffed when it is sent, or is simply buggy and does not do the byte stuffing correctly.

**Comment by klutvott123** [comment]
@mikeller Some more details which i should have added: 
When setting a value in the lua script to 125 or 126 and saving, it get's sent correctly to the FC. It's when the script tries to update the page again that it fails and all fields are blank. When i check the configurator i can see the value that i just set in the script. So it's working when sending data from transmitter but not when receiving.

You may be right about the lua API. I just had a quick look at some of the opentx s.port code and it looks like it's unstuffed correctly. Will investigate some more.

**Comment by bhuism** [comment]
+1

**Comment by klutvott123** [comment]
@mikeller So i have done some testing and made some observations. 

I soldered a UART module directly to the f.port pad on my FC and printed the data while the lua script was sending the pid command 112. I set my yaw I gain to 125 so that the frame had to be bytestuffed. The data for the pids is split up into four frames, is bytestuffed correctly and the CRCs are correct so no problems there. Just had to make sure. 

Then i modified the function that processes frsky telemetry data to check for the byte sequence i'm interested in and print that frame directly to the transmitter lcd, bypassing lua completely. Works fine for the frames that aren't bytestuffed but for the bytestuffed frames i get nothing. This makes me think that the bytestuffed frames aren't received at all. 

I had my transmitter turned on for a while and suddenly got a "sensor lost". Looked at the telemetry screen and saw the fuel sensor blinking with the value 124. It was gone until the fuel sensor reached 127. So the sensor got lost when the fuel sensor frames were bytestuffed. 

I also think i know why the lua pages are updating so slowly. For uplink data frames the CRC is calculated from all the bytes in the frame. APPID bytes included where the first byte is the frameID that isn't the same everytime. Sometimes the frameID causes the CRC to become 125 or 126 and then the frame has to be bytestuffed. The lua script then has to request the data again and again until none of the data or CRCs of the frames are bytestuffed. 

Can also confirm that s.port works correctly.

**Comment by RipperDrone** [comment]
Wow, @klutvott123 - you have finally found the explanation for what has been plaguing so many users for a long time. Wonderful work, thank you! Now it seems about how to do bytestuffing / destuffing the correct way, without the influence of channel values 125, 126...

Will this have to be done with FrSky devs, or is it something on the BF code side?

**Comment by bhuism** [comment]
@klutvott123 awesome

**Comment by klutvott123** [comment]
@RipperDrone I'm not 100% sure yet. I'll have to do some more investigating. 

Just did some testing with https://github.com/betaflight/betaflight/blob/9ed2914489fc12e60bcb1d8af1bdb65677518aa2/src/main/telemetry/smartport.c#L274 where i commented out the if else statement that takes care of bytestuffing on the FC side. Then i added serialWrite(port, c); below so that no bytestuffing is done on the data sent from FC to RX. 
Still using my pids for testing this and when i set any value to 126 it works. I get the correct sequence of bytes in opentx. The byte that comes after the pid value that is set to 126 is also shown correctly. Still doesn't work with 125. 

@mikeller is there a chance that the f.port firmware expects to do the bytestuffing itself? The length of an uplink dataframe is 8 bytes according to the documentation. When bytestuffing is done on the FC side the frame length exceeds those 8 bytes.

**Comment by RipperDrone** [comment]
@mikeller Is there anybody from FrSky active on BF GitHub so we can tie that rep/dev into the coding himself/herself? Reverse engineering the protocol details by @klutvott123 might be a tedious task to do...

**Comment by RipperDrone** [comment]
Sry for being too unexperienced to really comment in a knowledgeable and helpful way, but what I understood so far is there seems to be a mismatch btw sender and receiver regarding bytestuffing logics. Actual values of 125 (0x7D) seem to be interpreted wrongly as a bytestuffing yes/no flag, so the value itself is not transmitted properly anymore. I skimmed the telemetry.c code for occurrences of the FSS_DLE (=0x7D) flag and found several places where the value itself is still transmitted, but e.g. in this one, it isn't:
![image](https://user-images.githubusercontent.com/12601105/44618449-4ca09600-a876-11e8-8e8e-f3ef56a38b79.png)

Anyway, we desperately need a FrSky dev here to look into the receiver side / their protocol details to resolve it quickly... :-/

**Comment by klutvott123** [comment]
@RipperDrone That code is for receiving s.port/f.port data and it works correctly. In betaflight f.port is implemented as it says it should be in the f.port documentation. The only thing is that in the documentation it says bytestuffing should be done by "sender". Sender could be interpreted both as FC and RX as they are both senders. 

Heres some example data from my FC f.port pad:
```
126 8 1 48 53 0 112 112 0 0 176 126 <-- MSP_PID 112 from lua script

8 129 50 23 15 46 45 25 50 119         <-- First frame sent by FC containing PIDs

126 25 0 228 0 31 248 196 7 12 96 0 3 24 0 0 0 0 0 0 0 0 0 0 0 0 84 66 126

126 8 1 0 0 0 0 0 0 0 246 126

126 25 0 228 0 31 248 196 7 12 96 0 3 24 0 0 0 0 0 0 0 0 0 0 0 0 84 66 126

126 8 1 16 0 0 0 0 0 0 230 126

8 129 50 8 51 27 40 125 93 0 72   <--Second frame sent by FC. 125 is yaw I gain. Here it's bytestuffed as 125 93

```

**Comment by klutvott123** [comment]
I think i can confirm that there's bytestuffing going on in frskys f.port firmware. 
Set my yaw I gain to 126. The databytes sent from FC to RX are:
```
51 27 40 126
```
The databytes that comes out in opentx before any processing is:
```
51 27 40 125 94
```
This is with the codechanges i described above that disables any bytestuffing of data to be sent in betaflight. Still no idea why it doesn't do it when the frame contains 125.

**Comment by bhuism** [comment]
I love it, feel a work around is around the corner, no pun intended 😁

**Comment by mikeller** [comment]
@klutvott123: I did a couple of tests with r-XSR and XSR, and they both seem to be handling this in the same way, which is weird since they do not share the codebase, making it unlikely to be an unintentional bug.

Can you run the tests above to see what 125 comes out as?

@RipperDrone: Unfortunately most of the FrSky engineers do not speak English, making communicating with them only possible via proxy. But I am working my contacts to try and get them to look into this, and comment on it. Also, there are different teams responsible for the firmwares for XSR, r-XSR, and R9, making this even harder.

**Comment by mikeller** [comment]
Update: There is a difference: Disabling byte-stuffing will not fix 126 on XSR. And I am talking to FrSky, they will be looking into it.

**Comment by mikeller** [comment]
@klutvott123: You seem to be set up to do in / out comparisons. Can I get you to provide some in / out samples for byte-stuffed versions of 125 and 126 (i.e. `125 93` and `125 94`)? According to FrSky, the specification has to be interpreted to do the byte-stuffing in the same way as it is done in SmartPort, i.e. the flight controller has to do the bytes-stuffing.

**Comment by klutvott123** [comment]
@mikeller Ok, so i changed the smartPortSendByte function back so that it bytestuffs as usual. The thing is that when data from the FC has 125 or 126 in it, it doesn't show up at the opentx end at all so there's nothing to compare to. It turns out that enabling telemetry logging when compiling opentx makes it log everything. Couldn't find the frames containing 125 or 126 there either. 

Here's an example from the log: 
```
2018-08-26,12:27:29.870: 7E 00 32 16 0F 2E 2D 19 32 02   <-- First frame after sending MSP_PID
2018-08-26,12:27:29.880: 7E 00 32 08 35 32 4B 28 00 EA   <-- Third frame
2018-08-26,12:27:29.900: 7E 00 32 09 00 4D 00 00 00 77   <--Fourth frame

```
Second frame contains 125 or 126 and is bytestuffed in the FC. This pattern repeats itself through the log. 

Another example from the log: 

```
2018-08-26,12:27:32.150: 7E 00 32 02 00 4D 00 00 00 7D 5E
```
This frame contains the CRC for the MSP frame itself and its 0x4D. The CRC for the f.port frame at the end is 0x7E and has been bytestuffed, BUT it has to have been bytestuffed by the RX since the CRC would be different for the frame coming from the FC since instead of 0x00 at the beginning it would have been 0x08 and 0x81 and the CRC would have been calculated from that. So the RX takes the data from the FC, extracts it and repacks it. 

One last example:
```
2018-08-26,12:26:48.060: 7E 00 32 01 00 4F 00 00 00 7D 5D
```
This also contains CRC for MSP frame. The RX has calculated a new f.port CRC of 125 and bytestuffed it. 

Even if the frames from the FC containing 126 were to be passed on to opentx they would have been double stuffed by the RX. At least for the R-XSR since your testing shows that it doesn't do it on the XSR. 

Also thank you for contacting frsky about this!

**Comment by mikeller** [comment]
Thanks for the testing @klutvott123. From the look of it, the RX discards all frames that contain an 125 - this is definitely a serious problem.
The byte-stuffing of the 126 in itself is not a problem, since 126 should never be received on the RX side. Also, this can be explained with the fact that the byte-stuffing is only used on the serial links on both sides, but not on the radio link (i.e. an 126 is always sent as an 126 over the radio).

**Comment by bhuism** [comment]
I will test later today, thanks @mikeller

**Comment by RipperDrone** [comment]
@mikeller Did a thorough test on Ummagawd / Omnibus F4SD / R-XSR / BF 4.0 #1082 / FPORT / Taranis X9D+ OTX2.2

FrSky FPORT firmware 180615_Beta:
- When I set a PID value of 125 or 126, it does not read back these values on the RC - shows empty fields only
- Telemetry page Betaflight / PWM completely empty, LUA script on Taranis never successfully reads any content

FrSky FPORT firmware 180827_Beta (which you provided from FrSky):
- both 125 and 126 values are correctly read back :-)
- Betaflight / PWM page readout successful! :-)

Still not resolved (but minor glitch): Whenever I write back altered values over Lua Script, the Taranis screen says 'Saving', then 'Retrying'. Values are still stored correctly.

Summary: BIG WOW! We finally got this nasty bug resolved - R-XSR FPORT is back in the game! Loving it, great team effort and super smart finding by @klutvott123 👍

**Comment by Billj747** [comment]
@RipperDrone - FrSky FPORT firmware 180827_Beta solves the LUA script / "Sensor Lost" problem?

Where can I find that software version?  All I see under the frsky 'beta' firmware page is 180615.

**Comment by mikeller** [comment]
@Billj747: Here: https://github.com/betaflight/betaflight-tx-lua-scripts/issues/151#issuecomment-416183224

**Comment by mikeller** [comment]
@RipperDrone: I suspect that the timing used in the lua script is a bit too tight - we'll probably have to extend the time until we show 'retrying'...

**Comment by RipperDrone** [comment]
Nice, this is cream on top 👍 Happy to test when needed.

**Comment by bhuism** [comment]
@mikeller, @klutvott123, Just flown 4 packs with xsr-r on FPORT firmware LBT180827_Beta, BF 3.5.0 on kakute f7, OpenTX 2.2.2 on an X9D, had **0** 'sensor lost' warnings! very happy with this fix and collaboration, thanks all!

**Comment by mikeller** [comment]
@bhuism: Glad to hear that this seems to be fixing these sporadic and hard to pinpoint 'sensor lost' warnings as well - my suspicion is that they occurred whenever a sensor value happened to contain a 125 or 126.

**Comment by klutvott123** [comment]
Extending the save timeout makes no difference. I extended it to 10 seconds and it just "saves" for 10 seconds. Then "retrying" every time. I think the data is lost somewhere after being sent. When loading pages there are often frames lost. Sometimes one frame and sometimes all frames(this might also be that the request itself is lost).

**Comment by mikeller** [comment]
Ok, then we probably need to add multiple retries before we show 'retrying', or just accept that it won't work reliably first time round in most cases - the way telemetry data is framed for transport makes the link very unreliable for our purposes.

**Comment by RipperDrone** [comment]
It's so annoying, plus confusing for beginners. I'd go with 3x retrying w/o throwing the warning, only once it has to retry 4+ times, throw the message.

Alternatively, it could be a certain timeout. Guess implementing a counter is easier...

**Comment by klutvott123** [comment]
I just made a discovery. Most of the time the values will actually be saved on the first try, but it goes through all the retries anyway for some reason. I tried increasing the number of retries and it just goes through all of them it seems. 

If anyone has the time, try this. Set this line: https://github.com/betaflight/betaflight-tx-lua-scripts/blob/523fa04df1075639cd0fb4c9531171475fddae20/src/SCRIPTS/BF/protocols.lua#L12
to 0.

This allows it to save only 1 time. Will investigate more later.

**Comment by SomeguyinPDX** [comment]
Good morning to all.  If this is the wrong place please correct me.   I have been following this thread hoping it may have some answers to my R-XSR Lua script issues.   I'm am a brand new convert from Spektrum & a computer idiot so 95% of this thread is Greek to me.   I gave up trying to get Fport to work with the XRotor F4 45A ESC combo so I am currently wired with the non-inverted S.port pad from the R-XSR.    Lua scripts are showing telemetry data with the occasional "sensor lost" warning but no changes to the PIDS and VTX actually take place on the FC even though the X9D shows the change in the Lua script page until power is cycled then the telemetry updates and shows the original settings. Although if I use the BF OSD/stick menu I can make the same PID/VTX changes and they actually change.

**Comment by klutvott123** [comment]
@RipperDrone that sounds weird. Did you try rebooting both transmitter and FC before flashing to 180615? I had something similar happen once where i first got "sensor" lost, then "telemetry lost". RC controls didn't work either. Rebooted transmitter and FC and everything was fine.

**Comment by RipperDrone** [comment]
@klutvott123 Yes, did that multiple times. Can be reproduced in 100% of the cases. Also tried to switch on FC and RC in different order, no change. Changed distance RC - FC as well.

FPORT radio connection always stays intact, it's just telemetry/lua sensor information dying...

**Comment by klutvott123** [comment]
@RipperDrone Sound really strange as i would assume all betaflight targets would do the communication in exactly the same way since they're all using the same code. F.port is setup exactly the same as on your other F4 boards right? Tried any other tx pins? Sure the 180827 firmware haven't been corrupted somehow? Re-download it? I really don't know why that is happening, but i would try everything i can think of. Double check and triple check etc. 

@SomeguyinPDX not really the right place, but at the same time the right place since this thread has a f.port firmware that fixes a problem that you will encounter sooner or later. I would flash the R-XSR with the 180615 beta firmware first, then get it working, and then flash the 180827 firmware that you can find further up in this thread. For setting up f.port i would start with this: https://oscarliang.com/setup-frsky-fport/. Play around with the serialrx_inverted, serialrx_halfduplex.

**Comment by RipperDrone** [comment]
@klutvott123 
- Flashed the Ummagawd F4SD from the same file on my Taranis, it isn't corrupt. Retried flashing, dying sensors are remaining to be an issue.
- our remaining minor issue: Trying to understand the code regarding 'retrying': It starts with  setting currentState to pageStatus.Saving and zeroing saveRetries, at the same time initializing saveMaxRetries to 2, saveTimeout to 1.5s:

   if currentState == pageStatus.saving then
            saveRetries = saveRetries + 1
        else
            currentState = pageStatus.saving
            saveRetries = 0
            saveMaxRetries = protocol.saveMaxRetries or 2 -- default 2
            saveTimeout = protocol.saveTimeout or 150     -- default 1.5s
        end

- Then pageStatus.Saving leads into the if branch to check for timeout and exceeded # of save attempts, in this case calls saveSettings() again:

function run_ui(event)
    local now = getTime()
    -- if lastRunTS old than 500ms
    if lastRunTS + 50 < now then
        invalidatePages()
    end
    lastRunTS = now
    if (currentState == pageStatus.saving) then
        if (saveTS + saveTimeout < now) then
            if saveRetries < saveMaxRetries then
                saveSettings()
            else
                -- max retries reached
                currentState = pageStatus.display
                invalidatePages()
            end
        end
    end

- In saveSettings, saveRetries is incremented from 0 to 1:

if currentState == pageStatus.saving then
            saveRetries = saveRetries + 1

- it then falls through down to text display function where 'Retrying' is displayed once saveRetries is >0:

 elseif currentState == pageStatus.saving then
        lcd.drawFilledRectangle(SaveBox.x,SaveBox.y,SaveBox.w,SaveBox.h,backgroundFill)
        lcd.drawRectangle(SaveBox.x,SaveBox.y,SaveBox.w,SaveBox.h,SOLID)
        if saveRetries <= 0 then
            lcd.drawText(SaveBox.x+SaveBox.x_offset,SaveBox.y+SaveBox.h_offset,"Saving...",DBLSIZE + BLINK + (globalTextOptions))
        else
            lcd.drawText(SaveBox.x+SaveBox.x_offset,SaveBox.y+SaveBox.h_offset,"Retrying",DBLSIZE + (globalTextOptions))
        end
    end


- Therefore, it should have to do with the precondition to branch back into saveSettings, thus the watchdog timer must trigger the re-entry and the 'Retrying' display as a consequence...

saveTS + saveTimeout < now

Can't fully understand the timer details on my end :-/

**Comment by RipperDrone** [comment]
- Re timers: Is now counting forward or backwards? It says saving will happen as long as saveTS+saveTimeout are SMALLER than now. Is this correct?
- units: 150 value corresponds to 1.5s - is this correct?

     if (saveTS + saveTimeout < now) then
            if saveRetries < saveMaxRetries then
                saveSettings()
            else
                -- max retries reached
                currentState = pageStatus.display
                invalidatePages()
            end
        end

**Comment by klutvott123** [comment]
saveTS is the time where it started saving. If saveTS + saveTimeout is less than the current time it means that at least saveTimeout has passed since it started saving. Look at it this way: if now(current time) is larger than saveTS(time where saving started) + saveTimeout, then AT LEAST saveTimeout must have passed. 

Yes, 150 corresponds to 1.5s, but the timeout used for s.port/f.port is 300(3 seconds). 

The way i see it right now is that the whole "saving" and "retrying" is mostly for show. I don't see how it can know if the values have been saved without getting some kind of response from the FC. I'll look further into this tomorrow. 

Just for fun/testing, try this: 
After this line: https://github.com/betaflight/betaflight-tx-lua-scripts/blob/523fa04df1075639cd0fb4c9531171475fddae20/src/SCRIPTS/BF/ui.lua#L283
Paste this:

```
if currentState == pageStatus.saving and mspProcessTxQ() then
	currentState = pageStatus.display
	invalidatePages()
end
```
Try saving now 😄

**Comment by RipperDrone** [comment]
@klutvott123 @mikeller 

Tried it - saves fine, no more messages. Even no Saving... message displayed (maybe just a fraction of a second? Cannot read it, though). Then empty fields are shown, followed by reading the entries back from the FC which takes 2-3secs. Finally, the updated values are displayed.

So it's light and shadow: Retrying...  message is suppressed, but GUI is very un-intuitive to the user now: Not hinting at what's going on while empty fields are being shown...

It looks like the old implementation was showing the Saving... message for quite a long time, maybe covering the read-back period as well?

**Comment by RipperDrone** [comment]
@klutvott123 - as far as your statement 'just for show message displays, cannot know when saving was successful': Can't the system read back the altered values from the FC and compare against the RC/Lua inputs, thus confirming they have successfully been stored, page by page?

**Comment by RipperDrone** [comment]
- hm, tried to set saveRetries in message branching to 1 instead of 0, hoping saveRetries would then never make the way to the 'Retrying' message branch, but it does. So it seems that the saveRetries counter will increment up to 2 in any case!??:

if saveRetries <= **1** then
            lcd.drawText(SaveBox.x+SaveBox.x_offset,SaveBox.y+SaveBox.h_offset,"Saving...",DBLSIZE + BLINK + (globalTextOptions))
        else
            lcd.drawText(SaveBox.x+SaveBox.x_offset,SaveBox.y+SaveBox.h_offset,"Retrying",DBLSIZE + (globalTextOptions))
        end

**Comment by RipperDrone** [comment]
- SaveSettings() really seems to increment saveRetries regardless of whether saving was successful or not...

local function saveSettings(new)
    if Page.values then
        if Page.preSave then
            payload = Page.preSave(Page)
        else
            payload = {}
            for i=1,(Page.outputBytes or #Page.values) do
                payload[i] = Page.values[i]
            end
        end
        protocol.mspWrite(Page.write, payload)
        saveTS = getTime()
        if currentState == pageStatus.saving then
            saveRetries = saveRetries + 1
        else
            currentState = pageStatus.saving
            saveRetries = 0
            saveMaxRetries = protocol.saveMaxRetries or 2 -- default 2
            saveTimeout = protocol.saveTimeout or 150     -- default 1.5s
        end
    end
end

**Comment by klutvott123** [comment]
@RipperDrone The code i posted was WRONG. It was supposed to be: 
```
if currentState == pageStatus.saving and  not mspProcessTxQ() then
	currentState = pageStatus.display
	invalidatePages()
end
```
Can you try this? 

mspProcessTxQ() returns "false" when all packets in the transmit queue has been sent. This means that it will only show the saving message as long as there's still data to be sent. If it takes to long, it will retry as usual.

**Comment by RipperDrone** [comment]
Nice - Shows no more 'Retrying' now! Tried to change vlues on each page, message 'Saving...' comes up for ~1-2secs, then showing blank fields, then showing read back values. Reading back someimes takes a while, maximum around 3-4 seconds, but that's fine. We get rid of the 'Retrying' message finally :-)

**Comment by klutvott123** [comment]
@RipperDrone I think after sending the values to be saved, the script then waits for a response here: https://github.com/betaflight/betaflight-tx-lua-scripts/blob/523fa04df1075639cd0fb4c9531171475fddae20/src/SCRIPTS/BF/ui.lua#L363

If it receives the response it wants before the save timeout has elapsed, it invalidates the pages and they will reload with the new values. If no response is received before it timeouts, it will retry. It works just fine for s.port but not for f.port. I suspect that maybe no response is sent when using f.port, causing it to go through all the retries. If no response is sent, it has to be something on the betaflight side of things. The code i posted above just invalidates the pages when all the data is sent and doesn't wait for a response. 

@mikeller any thoughts on this?

Also i think the requestTimeout of 800ms is little too long for f.port. Sometimes it timeouts just when the data is about to arrive. Try setting it to 100(1 second). The downside is that if it does timeout, you have to wait even longer.

**Comment by klutvott123** [comment]
I have soldered my uart adapter back on to my f.port pad. When pressing save i can see all the frames that were sent from the lua script arriving at the FC. Right after the last frame is received, the FC sends back the MSP_SET_PID command like this:
```
8 1 48 33 0 0 139 0 0 26 126 <---last pid frame containing CRC
8 129 50 29 0 202 0 0 0 92 126 <---write command sent back from FC
```
This makes sense since the lua script expects the write command to be sent back here: https://github.com/betaflight/betaflight-tx-lua-scripts/blob/523fa04df1075639cd0fb4c9531171475fddae20/src/SCRIPTS/BF/ui.lua#L95

When the write command has been received back from the FC, the script sends the eeprom write command back to the FC. This command arrives at the f.port pad like this: 
```
8 1 48 50 0 250 250 0 0 158 126
```
The script then expects to get that command sent back from the FC here: https://github.com/betaflight/betaflight-tx-lua-scripts/blob/523fa04df1075639cd0fb4c9531171475fddae20/src/SCRIPTS/BF/ui.lua#L104
The thing is that the eeprom write command that the FC is supposed to send back i nowhere to be seen. It is never sent back. I can confirm this by invalidating the pages after receiving the write command back. Then it works just like with s.port. If anyone wants to try it, do this: 
After this line https://github.com/betaflight/betaflight-tx-lua-scripts/blob/523fa04df1075639cd0fb4c9531171475fddae20/src/SCRIPTS/BF/ui.lua#L101
Add this:
```
InvalidatePages()
```
This is just for testing, but i think we're closing in on a proper fix for this.

**Comment by RipperDrone** [comment]
Whoa, this sounds a bit too complicated to me - so you imply that the sequence of triggering an eeprom write command and acknowledging it back from FC to RC is coded differently in BF firmware for FPORT than it is coded for SBUS? In this case, the bug can be found in BF smartport.c / smartport.h?

**Comment by klutvott123** [comment]
It's not really that complicated. First you send all the frames containing the values you want to save. The FC responds with the write command for that msp frame to let you know that it received all the data correctly. When you get that response, you send the eeprom write command, telling the FC to save the values. The FC sends the eeprom write command back again so that you know that it actually saved the values. You then send a reboot command to the FC if the current page calls for that(PWM page needs that). 

The eeprom write command isn't sent back from the FC, and that's where this whole thing fails. 

I suspect fport.c might be the place, but i think @mikeller might have a pretty good idea about where to look 😄

**Comment by mikeller** [comment]
lol, this is ironic:

https://github.com/betaflight/betaflight/blob/adc965327f288d23a840077aab3e18d9ba3296f5/src/main/fc/config.c#L533-L537

**Comment by mikeller** [comment]
@klutvott123: Do you want to test betaflight/betaflight#6688?

**Comment by klutvott123** [comment]
@mikeller just tested it and it works. Thank you!
After the script sends the eeprom write command, this is sent from the FC: 
```
8 129 50 19 0 250 0 0 0 54
```
It took about 120ms for it to be sent back from the FC. But i think this can be adjusted for by increasing the timeout delay like you suggested earlier. I still find it strange that s.port has been working the whole time. Wouldn't frames be skipped for s.port too?

**Comment by mikeller** [comment]
@klutvott123: Thanks for testing, glad to hear we were able to stick a fork into this! The difference between SmartPort and FPort is that SmartPort is a telemetry protocol, and as such is not suspended during EEPROM operations. FPort on the other hand is an RX protocol (with integrated telemetry), and is suspended.

**Comment by klutvott123** [comment]
@mikeller Thanks for explaining. That makes sense. I didn't think of that. 
I increased the save timeout to 5 seconds, and now it saves on the first try most of the time. The shorter timeout delay causes the PWM page to sometimes go through all the retries. The values are saved but it doesn't do the reboot like it's supposed to. Looks like 5 seconds is just enough to make it work. I'll do some experimenting with timeouts/retries. It does make sense that f.port has to retry a bit more than s.port since we know that frames sometimes go missing even if they are sent from the FC.

**Comment by mikeller** [comment]
@klutvott123: I have put some more thought into this, and the only reason I can see for this to work better with SmartPort than with FPort is the lower data rate of SmartPort, which probably prevents UART buffer overruns while the CPU is blocked with the EEPROM write.

But the whole suspension is mumbo jumbo, since we do not use preemptive multitasking, and as a consequence there is no way that the code checking the suspension **during** it will ever be run, only the code checking for the frames to skip after the EEPROM write / read function has returned.

I have entirely removed the suspension for serial RX, which might further improve the FPort saving situation, please test.

I'll open another pull request for the suspension to be removed for PWM / PPM as well, and only leave the skips after the fact. But this should go into 4.0, since it needs some testing.

**Comment by RipperDrone** [comment]
@mikeller @klutvott123 Thank you so much, guys. It's sooo freakin' satisfying we successfully found the (2!) code lines responsible the bug :-) It was worth the effort, thank you @mikeller for interfacing with FrSky so effectively, and thank you @klutvott123 for sticking into it so much - great pleasure!

@mikeller - should I close this issue item now?

**Comment by bhuism** [comment]
Same here, thanks @mikkeller @klutvott33 @RipperDrone, and others, great to see this particular one fixed

**Comment by klutvott123** [comment]
@RipperDrone Please don't close yet. I have more info on this. 

@mikeller I actually got the 3.5.0 build i was on to work without the changes you made. 

To test your changes i had to make a 4.0.0 build to get it working because of some changes that have been made in feature.c/feature.h. It worked, but with no led strip. I see you made a pull request to remove led_strip support for F3 targets during 4.0.0 development. I disabled some features and reenabled the led_strip and the saving stopped working. Disable the led strip makes it work again. 

I then implemented the changes you made in the latest 3.5.x code and tested with that. Led strip is enabled by default and it doesn't work. Goes through all retries. I disabled the led strip in the 3.5.x build and it started working again. 

Then i went back to the 3.5.0 build i was using before, and disabled the led strip, and it works. This is without the changes in config.c/rx.h/rx.c.

**Comment by klutvott123** [comment]
Update: I reenabled the led strip on my 3.5.0 build(still without https://github.com/betaflight/betaflight/pull/6688). I then disabled a lot of other features i don't need, and now it works with led strip enabled, so it doesn't have anything to do with the led strip feature itself. This brings down the FLASH/RAM usage dramatically. Might it have something to do with that?

**Comment by klutvott123** [comment]
Ok, i think i understand this now. Just for testing i disabled writing the config to eeprom in rx.c. When "saving" from the lua script, it works as good as everytime on the first try(except that nothing is actually saved of course). 

@mikeller what you said earlier about uart buffer overruns happening during eeprom writes explains this. I assume that betaflight writes the entire config to eeprom. This would explain why disabling features helps with this, since it would have to write a lot more with a lot of features enabled. 

I think i'm satisfied now.  🕺

**Comment by RipperDrone** [comment]
Alrighty, closing then 👍

**Comment by RRAlex** [comment]
So for the layman, should we:

- use the beta firmware posted above?
- update the LUA scripts?

... Or wait for some upcoming new release?
Thanks :)

**Comment by klutvott123** [comment]
@RRAlex The beta firmware posted here fixes the issue where it doesn't send data containing the values 125 and 126, causing the pages to go blank, so you might as well give it a try. It also fixes the "sensor lost" issues. 👍

**Comment by mikeller** [comment]
@klutvott123: Thanks for the testing. And yes, the number of enabled features (at build time) will be proportional to the length of time that the firmware locks up during EEPROM writes / reads, and therefore to the probability of dropped frames / buffer overruns - not much we can do there without adding a whole new layer of complexity.

@RRAlex: betaflight/betaflight#6688, which will improve this, is scheduled to go into 3.5.1, due in the next few days.

**Comment by mikeller** [comment]
@klutvott123: Lol. There is a lot of gremlins in this code, apparently. I'll talk to FrSky again.

**Comment by paulo-serrao** [comment]
It would be great if you guys can figure out the fix for smartport as well, I always get the "Retrying" error while trying to save some combinations of VTX Channel/Band/Power, while on others it always work fine, I reported it here in the beginning of the year: https://github.com/betaflight/betaflight-tx-lua-scripts/issues/112

**Comment by klutvott123** [comment]
@paulo-serrao It's up to frsky to fix this now. We have found out what's wrong with both smartport and f.port now. To confirm that you're experiencing the same issues, try setting any value on any page to 125 or 126 and save. Goes through all retries and doesn't save? Then that's probably what's going on when you're trying to save your vtx settings. Some combinations will make the checksum for the MSP frame 125/126 and it won't save.

Also what firmware version are you using? Same problem with all of them? R-XSR receiver?

**Comment by klutvott123** [comment]
@paulo-serrao sounds about right. It's a RX firmware issue so it should be independent of which flight controller you're using.

**Comment by atomiix** [comment]
@mikeller Frsky just released the new beta firmware on their website. I was hopping that they would do the same for the x4rsb but it seems that no. Do you know if it's planed ?

**Comment by cgaxtreme10** [comment]
> set gyro_calib_duration = 128

Actualicé el firmware a BF 3.5.1 y persistia el problema: no cargabala página PWM en lua script de Taranis QX7, hice el cambio en CLI de set gyro_calib_duration a 128 y ahora funciona perfectamente. Muchas gracias!

**Comment by paulo-serrao** [comment]
@mikeller any idea if they have plans for "old" smartport (on any RX) as well?

**Comment by mikeller** [comment]
@paulo-serrao: I brought the fact that SmartPort seems to be having similar problems up with FrSky, and again I was told that they would look into it, but without a timeline.
I think we have to acknowledge there that FrSky is spread thin with a number of different hardware types used for their RX, and different firmwares for SmartPort and FPort, and will only be willing to allocate have a limited amount of engineering time to fixing and improving existing protocols, so this might take a while.

**Comment by mikeller** [comment]
@klutvott123, @paulo-serrao: FrSky has supplied a fixed version of the SmartPort r-XSR firmware that fixes the problem with byte-stuffed values (confirmed for the FCC version):

[RXSR_LBT180921_Beta.frk.zip](https://github.com/betaflight/betaflight-tx-lua-scripts/files/2405156/RXSR_LBT180921_Beta.frk.zip)
[RXSR_FCC180921_Beta.frk.zip](https://github.com/betaflight/betaflight-tx-lua-scripts/files/2405157/RXSR_FCC180921_Beta.frk.zip)

Please test and supply feedback.

**Comment by klutvott123** [comment]
@mikeller I can confirm that LBT version also works. Thanks! Good to see that frsky are fixing these problems.

**Comment by kalosjann** [comment]
I can confirm the FCC firmware provided by mikeller above is working on the following setup:
BF 3.5.1 OMNIBUSF4
HGLRC 20X20 STACK
R-XSR (by using the univerted s.port pin)
TBS UNIFY NANO

now I'm able to change PID as well as VTX channel

**Comment by RipperDrone** [comment]
Great FrSky was supporting on short notice by @mikeller to update their firmware. Other companies would have taken forever, I guess... Thank you all for resolving this one!

**Comment by paulo-serrao** [comment]
@mikeller It's also working on my R-XSR (using the univerted s.port pin).

I do get the lua script "out of memory" were it previously failed, but it does save (rebooting the Taranis shows that the new setting do stick)

**Comment by cexshun** [comment]
Beta firmware worked. However, the problem also occurs on the L9 r9mm. To my knowledge, there isn't a beta fix yet for the r9mm receiver.

**Comment by RipperDrone** [comment]
Beta FW verified on 2 more FPORT builds. Works nicely! Hopefully FrSky can apply a siilar fix to r9m, seems it needs the same fix...

**Comment by klutvott123** [comment]
Frsky released new flex firmware for all the r9 receivers a couple of weeks ago. Has anyone tried it? 

@RipperDrone yeah, xsr and x4rsb also need fixing. I'm just going to assume that ALL of the sbus/fport receivers have this issue and need fixing. Frsky has some work to do. Personally i'm not buying any frsky products until i start seeing fixes for receivers other than the r-xsr. 

IMO the r-xsr is the best option for anyone buying a new rx right now since it has known working firmware, but still there are issues with it. The EU-LBT(sbus and fport) firmware has some serious frame drop issues. Sometimes it's not too bad with only the occasional frames getting dropped, and sometimes it's dropping between one and four frames every 100ms. It's periodical and consistent throughout the whole flight independent of rssi. Seems like some sort of timing/sync/init issue. The "fix" for this is to bind the receiver without telemetry. I think what's going on is that the transmitter listens, finds that the rx is sending telemetry data on that channel and then doesn't send the frame. I have tested FCC firmware, and it doesn't have this issue with or without telemetry. @RipperDrone have you experienced this?

**Comment by RipperDrone** [comment]
@klutvott123 interesting, I've never experienced this since all my R-XSR are not flashed to LBT version... :-/

**Comment by klutvott123** [comment]
@RipperDrone I just assumed you used EU firmware since you're in Germany. Just keep using the FCC version. It performs much better 👍

**Comment by TheWiseWalrus** [comment]
@klutvott123  I have tried both at this point. Currently it is on F.Port, but it is easy enough to get it at to switch back. Later tonight, I'm going to see if the issue is that I'm using Tx2 for S/F Port (not at the same time, lol) and Rx2 for DShot. I will also try using a softserial pad (LED)

**Comment by TheWiseWalrus** [comment]
If you had to guess, would the above S.Port or the Beta F.Port give me a better chance of success? I only have 1 Tx pin broken out, so I have to use softserial for SmartAudio no matter what

**Comment by klutvott123** [comment]
@TheWiseWalrus the spracingf3evo manual says: "Note: When using a Serial RX receiver the TXD (T2) pin cannot be used for telemetry.  Use UART3 TXD 
instead.". Maybe using rx2 for dshot is a problem too. I would try fport with uart1 or 3.

The firmwares posted here won't help with setup problems, but i would recommend using them anyway since they fix an annoying bug that you will likely encounter sooner or later.

**Comment by TheWiseWalrus** [comment]
@klutvott123  Sorry for the confusion, I am using the Emax Femto F3 (manual: https://www.emaxmodel.com/download/F3_femto_manual_V1.2.pdf), which happens to use the same target. The above manual does recommend using Tx2 for telemetry and Rx3 for sbus, which is how I had it originally set up. I only have 3 uart pins (Rx3, Tx2, Rx2) I'll try unmapping motor 4 from Rx2 and see if that makes a change. After that I'll do the same with sport and then try softserial (which requires a custom hex I compiled, since it was removed to fit transponder in)

**Comment by klutvott123** [comment]
@TheWiseWalrus Somehow i missed that when reading your post. I don't see any reason for it not to work if you use uart2 for fport only. Make sure you have serialrx_halfduplex and serialrx_inverted set correctly.

**Comment by TheWiseWalrus** [comment]
@klutvott123 with it in s.port mode I tried all four combinations of tlm_... and it only showed the sensor data on my taranis with halfduplex and inverted on and off respectively. I also tried softserial, but that didn't seem to work with any combinations. I see the sensor data on my Taranis, so I am starting to think the issue is there, not with betaflight. No idea what settings to tweak there though . . .

**Comment by Pascal-m** [comment]
Hello @mikeller  do you have any news about R9MM RX firmware update?
To remind you of my problem it's the #151 issue that was raked here.
Thk

**Comment by mikeller** [comment]
@Pascal-m: No I nave not. I have notified FrSky that there is a problem, and received confirmation that they are looking into it, but not heard back since. I'll post updates here as soon as I get them.

**Comment by mikeller** [comment]
Hey XSR owners. Can you please test and provide feedback on this version (sorry, no LBT version yet):

[XSR_FPORT_FCC181018_Beta.zip](https://github.com/betaflight/betaflight-tx-lua-scripts/files/2561827/XSR_FPORT_FCC181018_Beta.zip)

This fixes the byte stuffing issues for me.

**Comment by mikeller** [comment]
Also (I can't test for lack of LBT hardware), can I please get feedback on

[XSR_FPORT_LBT181018_Beta.zip](https://github.com/betaflight/betaflight-tx-lua-scripts/files/2565106/XSR_FPORT_LBT181018_Beta.zip)

**Comment by mantasnakt** [comment]
I`m having same issue (lua scripts slowly refreshing and not saving values) with mine R9 MM and I`m ready to test any beta FW builds if there`ll be any

**Comment by RipperDrone** [comment]
should I update this issue's title, since it seems fixed for the R-XSR but still open for R9M and under testing for XSR? Maybe I'll generalize the title for now...

**Comment by hsuq510** [comment]
got the #186 and RXSR_FCC180921_Beta.frk.zip for smartport fixed it.

**Comment by bhuism** [comment]
@mikeller , I fly XSR_FPORT_LBT181018_Beta.zip this afternoon

**Comment by bhuism** [comment]
XSR_FPORT_LBT181018_Beta.zip flies great, and saves values without retrying

**Comment by klutvott123** [comment]
I have some new info on the FPORT saving situation. It seems to be temperature related. As you know, sometimes it works perfectly fine and sometimes it will happily retry as many times you allow it. At first this seemed to be random, but i have started monitoring the temperature of the stm32F405 on my ff fortinif4. When i first power it up it saves just fine and it gradually gets worse to the point of going through all retries. It reaches a temperature of 58C. Then i directed a fan directly to the FC and the temperature is steady at 40C. While keeping the mcu at this temperature it saves successfully at the first try most of the time, sometimes on second try and occasionally on the last retry. I have repeated this test several times and the same thing happens. 

I'm not sure, but i believe this might be caused by the internal RC oscillator drifting. Maybe too much drift for uart, but i don't think so. Max 4% at 85C according to stm32f405 datasheet. I've had uarts reliably  working in way worse conditions than that.  Maybe a combination of things like drift, pin slew rate etc. 

There is a resonator soldered right next to the chip on the ff fortinif4, but there's no ``TARGET_XTAL_MHZ`` defined for that target so i assume it's just using a generic config that uses the internal RC oscillator, but i might be wrong. 

@mikeller : Would be great if you could shed some light on this. Will the target use the external oscillator if i just define ``TARGET_XTAL_MHZ`` with the correct frequency and compile?

**Comment by mikeller** [comment]
@klutvott123: This sounds very weird - especially because if it was a drift problem with the UART, then surely the reception of RC data would be affected as well, which it isn't.

Betaflight always uses an external oscillator. If `TARGET_XTAL_MHZ` isn't defined it defaults to 8 MHz.

**Comment by vkolotov** [comment]
> got the #186 and RXSR_FCC180921_Beta.frk.zip for smartport fixed it.

Hey @hsuq510, where did you get it from? Can't find it anywhere...

**Comment by mikeller** [comment]
@vkolotov: https://github.com/betaflight/betaflight-tx-lua-scripts/issues/151#issuecomment-423505314

**Comment by vkolotov** [comment]
> @vkolotov: [#151 (comment)](https://github.com/betaflight/betaflight-tx-lua-scripts/issues/151#issuecomment-423505314)

Thank you @mikeller , I've tried it. Same thing, it displays only `---`. When I initiate a reboot for FC from lua script, after it reboots, it shows data from the current page, when I switch to next page all values are `---`. 

Also I noticed that it was working before a flashed that beta FW! It was stable, and I could change values. Then I flashed the beta and it stopped working again. I read somewhere that someone said it might be temperature related, I think he is right. When the RX is not hot it works fine, once it got hot it stops working...

**Comment by vkolotov** [comment]
Yes, it stopped working again once it got hot. Will power it up when it is cooled down and report back.

**Comment by vkolotov** [comment]
Yes, I can confirm this. When quad is cold, the lua script works fine.

**Comment by vkolotov** [comment]
> I have some new info on the FPORT saving situation. It seems to be temperature related. As you know, sometimes it works perfectly fine and sometimes it will happily retry as many times you allow it. At first this seemed to be random, but i have started monitoring the temperature of the stm32F405 on my ff fortinif4. When i first power it up it saves just fine and it gradually gets worse to the point of going through all retries. It reaches a temperature of 58C. Then i directed a fan directly to the FC and the temperature is steady at 40C. While keeping the mcu at this temperature it saves successfully at the first try most of the time, sometimes on second try and occasionally on the last retry. I have repeated this test several times and the same thing happens.
> 
> I'm not sure, but i believe this might be caused by the internal RC oscillator drifting. Maybe too much drift for uart, but i don't think so. Max 4% at 85C according to stm32f405 datasheet. I've had uarts reliably working in way worse conditions than that. Maybe a combination of things like drift, pin slew rate etc.
> 
> There is a resonator soldered right next to the chip on the ff fortinif4, but there's no `TARGET_XTAL_MHZ` defined for that target so i assume it's just using a generic config that uses the internal RC oscillator, but i might be wrong.
> 
> @mikeller : Would be great if you could shed some light on this. Will the target use the external oscillator if i just define `TARGET_XTAL_MHZ` with the correct frequency and compile?

I can confirm this, when my quad is cold, the telemetry works fine. When it get hot, it does not.

**Comment by klutvott123** [comment]
@mikeller Thanks for the clarification. Oscillator drift can be ruled out then. True about the RC data too. It's very weird indeed, but so far my experiments show that the successrate goes down when the temperature goes up. 

@vkolotov Sounds like you might be having other issues. If it's working with one firmware version, but not the other, i don't think it's caused by temperature. That it works when it's cold could also be because you just power cycled it. Which beta FW are we talking about? R-XSR or Betaflight itself? Maybe consider trying the latest stable release 3.5.3.

**Comment by vkolotov** [comment]
@klutvott123, when it is hot and not working no matter how many times you powercycle quad and TX it won't work. It is definitely temperature related.

**Comment by vkolotov** [comment]
I believe the problem is in the RX. I can do some experiments. I've got a disolder hot air gun and I think I an also use just a aircon, or a fridge :)?

**Comment by klutvott123** [comment]
Here's a follow up regarding my above comment about the retry situation being temperature related.
First here's some test data. The lua script is configured with 2 retries and 5 second timeout. I count a successful save if the ``eepromWrite`` command is received back within the allowed retries. 20 save attempts are made. 
```
Temperature = 40C
Successful saves = 17/20
Total retries = 8
```
Then i let it heat up:
```
Temperature = 58C
Successful saves = 11/20
Total retries = 27
```
Unfortunately i didn't log the data everytime i did this experiment, but the above results reflect the behaviour i'm seeing pretty well. 
I believe my statement about this being related to temperature and oscillator drift is correct, but it's not the direct cause. 

I set up LED1 to go out when saving to eeprom, and come back on when it's done. I would say it takes roughly 500ms(the exact time doesn't really matter). This means that the cpu is in a suspended state that entire time. Let's say the temperature drift of a typical crystal oscillator ranges between 10 and 30ppm. As an example let's assume the crystal temperature increases 10C and it drifts 5ppm. The write to eeprom should take the same amount of clock cycles every time, but with a 5ppm drift, 500ms becomes 499.75ms. A difference of 0.25ms. If it had drifted 20ppm the time difference would be 1ms. The drift is small enough that it doesn't cause any problems during normal operation, but in the special case of a write to eeprom it can. 

Have a look at this function: https://github.com/betaflight/betaflight/blob/3100264e8674f876d609256e9aee96e2d75c8afd/src/main/telemetry/smartport.c#L491.
This is the function that is called when we have a fport/sport telemetry request. If the request contains MSP data, it calls ``handleMspFrame``. Somewhere inside ``handleMspFrame`` a function calls a function which eventually saves to eeprom. It also prepares a response for us to send back. When ``handleMspFrame`` is done, the code immediately goes on to send a response. In the special case of a write to eeprom the cpu has just come out of the suspended state and is living 100s of ms in the past, and the initial telemetry request is long gone. The cpu doesn't know it and sends the reply blindly. The receiver might be listening again, or it might be busy sending us RC data. We can't know. This is where the oscillator drift comes in. That 0.25-1ms can easily skew things enough so that the cpu doesn't come out of suspended state exactly at the time when there's a new telemetry request. 

Fortunately there's an easy fix for this. Simply check if we received the MSP_EEPROM_WRITE command. If that's the case, skip a few telemetry requests to let everything get back to normal operation before sending the reply back. Here's what i did https://github.com/klutvott123/betaflight/commit/0b9105294ba6de261235382f4c47f0cd72f21716. 

If anyone wants to try this i suggest changing the smartport savetimeout in protocols.ui from 300 to 500. All my testing shows that with FPORT it very often takes more than 3 seconds, but rarely more than 5 seconds to save. I can provide you with a hex if you're not setup to compile betaflight. 

Here's some data. First i tested with the unmodified code:
```
Temperature = 54C
Successful saves = 10/20
Total retries = 31
```
Then with my fix:
```
Temperature = 54C
Successful saves 20/20
Total retries = 1
```
I have done multiple tests, and the results are consistent. The behaviour doesn't change with temperature anymore, and there's a dramatic reduction in retries. Data can of course still be lost in other ways.

**Comment by vkolotov** [comment]
Well done @klutvott123 , would this fix the issues with read operation as well?

**Comment by klutvott123** [comment]
@vkolotov Thank you 😃 . No it won't fix that. 
The script attempts to reload the page every 800ms. If it hasn't received all the requested data by then, it will retry again and the fields remain blank. As a troubleshooting step you could try increasing the request timeout. https://github.com/betaflight/betaflight-tx-lua-scripts/blob/f1b83637d597af7d0e3b5ea04f74330558ba7130/src/SCRIPTS/BF/ui.lua#L18
If this value is too low, you will never receive all the data before it retries. Try increasing it to 150 maybe and see what happens. Might not solve it, but worth a try.

**Comment by vkolotov** [comment]
Great, I'll try that on this weekend.

Is your fix/change going to be merged to the upstream soon?

**Comment by dejinco** [comment]
Cancel

Sent from a mobile device. Excuse brevity and typos.

On Dec 5, 2018 15:20, "Vlad Kolotov" <notifications@github.com> wrote:

Great, I'll try that on this weekend.

Is your fix/change going to be merged to the upstream soon?

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/betaflight/betaflight-tx-lua-scripts/issues/151#issuecomment-444673686>,
or mute the thread

**Comment by mikeller** [comment]
Wow, good finding @klutvott123!

I think your suggestion is a valid approach for a fix, I've added a couple of style remarks.

An alternative approach would be to change MSP_EEPROM_WRITE to be asynchronous, similar to how rebooting is done: https://github.com/betaflight/betaflight/blob/25e16072922e5cc238a7c480a74da93c6011d44a/src/main/interface/msp.c#L1449

But I think your approach is preferrable, since it only changes how MSP *over telemetry* works, and does not bring the risk of affecting the timing when used by configurator or other clients.

Do you want to open a pull request?

**Comment by klutvott123** [comment]
@mikeller Thanks! 😃 

I did consider the asynchronous approach, but decided against it for several reasons. I want this to apply for telemetry only like you said. Also i think it makes more sense if the request is executed before the reply is sent back. For rebooting the asynchronous approach makes more sense to me. Then theres a possibility for the FC to send a reply to the client before rebooting. 

I'm currently looking into your comments, and making/testing the changes. I will make a pull request when i'm happy with it. 👍

**Comment by ijustwannafly** [comment]
Fixing the saving/retry is great news as it is 1/2 the main problems at the moment.

Besides increasing the timeout for the read function what might help solve that issue?

After work I will test changing my timeout settings for the save function and see how that goes though I still have the occasional issue of the pages not loading/reading.

**Comment by klutvott123** [comment]
@ijustwannafly I believe it's caused by the amount of regular telemetry data that is sent. I did a test where i only allowed BF to send MSP data. Both loading and saving speeds are increased, but of course you get no telemetry. This gave a save time of about 600ms 😄 . I also saw someone made a pull request that allows disabling of frsky telemetry sensors, but i only had a quick look at it. One idea might be to implement some sort of telemetry throttling. For example only allow sending telemetry every other request, but still allow MSP at the highest rate possible.

**Comment by ijustwannafly** [comment]
> 
> 
> One idea might be to implement some sort of telemetry throttling. For example only allow sending telemetry every other request, but still allow MSP at the highest rate possible.

Along the lines of the msp write logic, could all other telemetry be "stopped" if an eeprom read (just guessing) is sent so that the telemetry feed is "free" of other data?

Theoretically if the lua script is reading the eeprom it shouldn't need to be fighting other telemetry data since most users wouldn't be using this particular lua script in flight.

Just a thought.

**Comment by klutvott123** [comment]
@ijustwannafly That's actually a good idea, but i think something like that already happens. The FC wouldn't know to stop the telemetry until it receives the command sent from the script. When BF receives an MSP command, it immediately starts sending the data back. MSP has priority over telemetry, so if there's still MSP packets to be sent, they are sent first. https://github.com/betaflight/betaflight/blob/25e16072922e5cc238a7c480a74da93c6011d44a/src/main/telemetry/smartport.c#L526

**Comment by ijustwannafly** [comment]
@klutvott123 

Maybe I'm getting some things confused here.

https://github.com/betaflight/betaflight-tx-lua-scripts/issues/177#issuecomment-429802946
>
>
>@ijustwannafly:

>>    Is the lua script really dependant on the sensors?

>mikeller says - Since the data is tunnelled through the RX / TX via a 'sensor': Yes it is.

One the one hand it is advised to discover sensors in order for the lua script to function properly.

This to me would suggest that telemetry data is not actually being stopped during this lua script use.

Happy to be corrected to the proper data/information.

**Comment by vkolotov** [comment]
If the number of sensors really makes the difference, it would be nice to have something similar to ibus telemetry where you can define what sensors you want to send: `ibus_sensor` setting.
https://github.com/betaflight/betaflight/blob/c400186e70f3da6ca8a1270d61b9d46ea616823a/src/main/interface/settings.c#L964

For example:
```
ibus_sensor = 1,2,3,4,0,0,0,0,0,0,0,0,0,0,0
```
I guess most of the pilot just need voltage and RSSI as telemetry, so why end gps and other stuff if it is not used...

**Comment by vkolotov** [comment]
> @vkolotov Thank you 😃 . No it won't fix that.
> The script attempts to reload the page every 800ms. If it hasn't received all the requested data by then, it will retry again and the fields remain blank. As a troubleshooting step you could try increasing the request timeout.
> 
> [betaflight-tx-lua-scripts/src/SCRIPTS/BF/ui.lua](https://github.com/betaflight/betaflight-tx-lua-scripts/blob/f1b83637d597af7d0e3b5ea04f74330558ba7130/src/SCRIPTS/BF/ui.lua#L18)
> 
> Line 18 in [f1b8363](/betaflight/betaflight-tx-lua-scripts/commit/f1b83637d597af7d0e3b5ea04f74330558ba7130)
> 
>  local requestTimeout = 80 -- 800ms request timeout 
> 
> If this value is too low, you will never receive all the data before it retries. Try increasing it to 150 maybe and see what happens. Might not solve it, but worth a try.

Sorry I'm not an expert in lua script. Do I just change it on my SD card? Do I need to rebuild the script? The documentation is a bit misleading saying that you should run a build script: https://github.com/betaflight/betaflight-tx-lua-scripts#building-from-source

**Comment by klutvott123** [comment]
@ijustwannafly Frames are sent with an ID which indicates which sensor the data is for. Different IDs for sensors and MSP. Even if MSP is just data, it's treated as a sensor. 

@vkolotov If you had ``.luac`` checked when you downloaded opentx, it will compile the ``.lua`` file again if you edit it directly on your SD card.

**Comment by ijustwannafly** [comment]
> 
> 
> @ijustwannafly Frames are sent with an ID which indicates which sensor the data is for. Different IDs for sensors and MSP. Even if MSP is just data, it's treated as a sensor.

I remember asking once somewhere why there isn't a bf sensor ID specifically for MSP traffic.

I think that would be a good solution for the communication between lua and bf.

**Comment by mikeller** [comment]
@ijustwannafly: If you have a look at https://github.com/betaflight/betaflight/blob/25e16072922e5cc238a7c480a74da93c6011d44a/src/main/telemetry/smartport.h#L58-L60 you'll notice that the sensor ID makes up 33% of the payload - hardcoding it to a fixed value will make MSP over SmartPort / FPort 33% slower, for no gain at all.

**Comment by mikeller** [comment]
@klutvott123: Not sure if throttling telemetry at all times just to get better UX while running scripts is the best way to do this - maybe we should make the throttling level configurable dynamically (over MSP), and have the scripts send a command to throttle down telemetry on entry, and throttle up again on exit?

**Comment by klutvott123** [comment]
@mikeller Good idea, i like it! Then the user could have it all without having to do anything. My only concern is that we could end up with a situation where the command is lost. It would have to be implemented with response, retries etc. Would it be a good idea to have ``ui.lua`` signal ``background.lua`` and have it send the throttling command? I'm thinking about just setting a variable when the run function starts running again, and clearing it in the protocols exit function. I might have a go at it later(but please feel free to do it if you want 😄), but right now i'm investigating the random page loading issues with fport/smartport. 

I have made some observations about this. I'm printing out the number of received frames and lost frames to see what's going on in real time. The script always starts receiving frames after sending a request, and retries when a frame is lost(sequence number doesn't match up). The funny thing is, when a frame is lost, it's ALWAYS after a multiple of 5 received frames. I'm testing with a custom LED page that needs 65 frames for all the data. Fails at 10, 15, 45, 55, 35, etc. Since i'm now seeing close to 100% success rate when saving, I believe that the same should be possible for page loading. Especially since MSP has precedense over regular telemetry in this case. It behaves like this even if i only allow MSP data. I believe there's a solution to be found. Just have find and explore all dead ends first, since that's usually how it goes.

**Comment by mikeller** [comment]
We could return the throttling status as part of the status call that we use in lua.

Re lost frames being a number of 5, this could be to do with the quite crude way that queues are implemented in the RX / TX, and long MSP frames resulting in queue overruns. I think the way it is currently, MSP is exempt from the back pressure / throttling that we are implementing for telemetry replies - maybe try changing that and throttling MSP as well, in order to avoid overflowing the queue on the RX - not sure if there's much we can do on the TX though.

**Comment by klutvott123** [comment]
@mikeller Sorry, my testing was a little off in my above post. I meant to check for lost frames, but counted bytes instead, and since a frame contains the same amount of databytes it would fail at a multiple of the amount of bytes.

I don't think there's any overruns occuring, and heres why. I'm logging the data received in opentx, and every time it retries loading the page, I see all the response frames arriving in opentx WITH the correct sequence numbers. There's something strange going on here, and i have some possibilities to explore. I'm wondering if opentx is stuffing telemetry frames into the lua input buffer at a fixed interval, and that we might end up with duplicate frames in there. For regular telemetry scripts it wouldn't matter, but for MSP it would break everything since it has to be in the correct sequence.

**Comment by mikeller** [comment]
Good finding, @klutvott123!

I suspect the answer to 'why' is 'because it can'. When telemetry packets are used for telemetry, they are inherently atomic and reentrant, so repeating / duplicating them is not an issue. Not sure if we can convince OpenTX to change this, in the past I've found them to be neither cooperative nor understanding of technical issues. :-|

**Comment by klutvott123** [comment]
@mikeller That's what i assume too. Doesn't matter for telemetry. Seems like our best bet is to implement a check for duplicate data and just discard it. I did a quick and dirty implementation just to test, and it gave a huge reduction in retries. Still not perfect since it doesn't account for problems that may arise because of it. At least now we know the cause. This is probably what a lot of smartport users have been struggling with lately too. I'll look further into possible solutions, but now i'm done. Been struggling with this all day 🤣

**Comment by vkolotov** [comment]
Good stuff guys, I'm listening to this thread and looking forward to test it.

**Comment by klutvott123** [comment]
Here's my attempt: https://github.com/klutvott123/betaflight-tx-lua-scripts/tree/discard-duplicate-data-from-buffer. I made a function to replace ``sportTelemetryPop()``. It returns the same except from the duplicates. This seems to be the most reliable way. I thought about just checking the sequence numbers, but that brings all sorts of other problems. Also i didn't want to start butchering the existing code. 
@vkolotov Test it if you want. Turn your accelerometer back on too 😃 

Keep in mind that this doesn't eliminate retries completely when loading pages. There are still the occasional lost packets. I logged all the data coming straight from the lua input buffer, and i see missing sequence numbers sometimes. I don't believe that it's because they were lost for real, but rather because opentx didn't manage to put them in the buffer. The behaviour I'm seeing in my logs indicate that data is put in the buffer at a variable rate. If the rate is too low, packets will be lost. Looks like the way this works is just not very suited for MSP data. 😏

**Comment by klutvott123** [comment]
Have been digging further into this. I have been studying the opentx code, and it looks fine to me. Data is put into the lua buffer when it receives new telemetry data. The problem is still that it contains duplicate data, but It's not opentx that is at fault. I apologize for saying that it was earlier. It's simply because opentx actually receives duplicate data. I looked closer at my previous raw telemetry log from opentx and it contains the duplicates too. For some reason i assumed that it was just sampling too fast, and reading the same data several times, but that's not the case. Opentx processes telemetry data as soon as the uart interrupt fires, and logs telemetry data at 10ms intervals. 

Just to be sure, I logged the data coming out from the FC too. No duplicates or missing packets there. So, I find nothing wrong with the data coming out of Betaflight, and nothing wrong with the way opentx handles telemetry. Perhaps this is just another frsky "feature"? 🤔

**Comment by mikeller** [comment]
FrSky X uses 2 bit sequence numbers for the telemetry frames sent over the air, which, as you can imagine, will break as soon as more than 3 frames are lost due to interference (and to make matters worse, one telemetry frame on the radio link is only 5 bytes long, so an entire SmartPort frame will not fit into a single frame. Looks like they are using another layer of error detection / retransmit in the SmartPort layer, creating duplicates in the process. :-|

**Comment by klutvott123** [comment]
What you're saying makes sense to me. In the cc2510 datasheet i find nothing about automatic transaction handling(which other RF modules like nrf24l01 offer), which indicates that it would have to be implemented by frsky themselves like you suggest. I imagine that it goes something like this: 
Send all packets containing the payload, wait for acknowledge. 
If no acknowledge, send payload again. 
Keep doing it until max retries(which could be 4 based on what the logs show) reached. 

It would only make sense if the XJT module passes the data along to opentx after performing the checks to determine if it's valid or not. To me this indicates that no acknowledge is received by the RX(or not received in time), and it retransmits the payload even if the data arrived successfully at the XJT the first time. This would explain the duplicates.

**Comment by mikeller** [comment]
@klutvott123: Correct, the frame loss detection / correction is done in the RX, not in the RF chip. To get an idea of how it works at the lowest level, have a look at `src/main/rx/cc2500_frsky_x.c` and its counterpart in https://github.com/pascallanger/DIY-Multiprotocol-TX-Module/blob/master/Multiprotocol/FrSkyX_cc2500.ino

**Comment by klutvott123** [comment]
@mikeller Thanks for the info! I'll have a look at those 👍
